src/psyclone/tests/domain/lfric/kernel/lfric_kernel_metadata_test.py

```diff
@@ -157,6 +157,721 @@ def test_init_args_error():
+    meta_args = [
+        ColumnwiseOperatorArgMetadata("gh_real", "gh_read", "w3", "w1"),
+        FieldArgMetadata("gh_real", "gh_read", "w2"),
+        FieldArgMetadata("gh_real", "gh_write", "w3")]
+    lfric_kernel_metadata = LFRicKernelMetadata(
+        operates_on="cell_column", meta_args=meta_args)
+    with pytest.raises(ParseError) as info:
+        lfric_kernel_metadata._validate_cma_apply_kernel()
+    assert ("In a CMA apply kernel, the function space of the read field "
+            "must match the function space of the CMA operator's 'from' "
+            "function space, but found 'w2' and 'w1' respectively in kernel "
+            "metadata 'None' for procedure 'None'." in str(info.value))
+
+    # OK.
+    meta_args = [
+        ColumnwiseOperatorArgMetadata("gh_real", "gh_read", "w3", "w1"),
+        FieldArgMetadata("gh_real", "gh_read", "w1"),
+        FieldArgMetadata("gh_real", "gh_write", "w3")]
+    lfric_kernel_metadata = LFRicKernelMetadata(
+        operates_on="cell_column", meta_args=meta_args)
+    lfric_kernel_metadata._validate_cma_apply_kernel()
+
+
+def test_validate_cma_matrix_kernel():
+    '''Test that the _validate_cma_matrix_matrix_kernel method behaves as
+    expected.
+
+    '''
+    # Check validate_cma_kernel is called.
+    lfric_kernel_metadata = LFRicKernelMetadata(operates_on="domain")
+    with pytest.raises(ParseError) as info:
+        lfric_kernel_metadata._validate_cma_matrix_matrix_kernel()
+    assert ("A CMA kernel should only operate on a 'cell_column'. However, "
+            "the kernel metadata 'None' for procedure 'None' operates on "
+            "'domain'." in str(info.value))
+
+    # CMA operators or scalars.
+    meta_args = [
+        ColumnwiseOperatorArgMetadata("gh_real", "gh_write", "w3", "w2"),
+        FieldArgMetadata("gh_real", "gh_write", "w3")]
+    lfric_kernel_metadata = LFRicKernelMetadata(
+        operates_on="cell_column", meta_args=meta_args)
+    with pytest.raises(ParseError) as info:
+        lfric_kernel_metadata._validate_cma_matrix_matrix_kernel()
+    assert ("A CMA matrix-matrix kernel must only contain CMA operators or "
+            "scalars, but found 'field' in kernel metadata 'None' for "
+            "procedure 'None'." in str(info.value))
+
+    # One CMA argument written to.
+    meta_args = [
+        ColumnwiseOperatorArgMetadata("gh_real", "gh_read", "w3", "w2")]
+    lfric_kernel_metadata = LFRicKernelMetadata(
+        operates_on="cell_column", meta_args=meta_args)
+    with pytest.raises(ParseError) as info:
+        lfric_kernel_metadata._validate_cma_matrix_matrix_kernel()
+    assert ("A CMA matrix-matrix kernel must write to one CMA operator "
+            "argument, but found 0 writers in kernel metadata 'None' for "
+            "procedure 'None'." in str(info.value))
+
+    # Other arguments must also be read only but remaining args will
+    # be scalars which must already be read only so no more checks
+    # required.
```

Comment:
While that's true, I think a test here wouldn't hurt in case something unexpectedly changes in the future. Please could you test that a kernel with a scalar that is written to is rejected.

Reply:
I've added a ScalarArgMetadata declaration test, which is what would raise the exception and extended the comment. Hopefully that is what you want.

Reply:
Thanks, I think that covers it.